kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableSchemaEvolutionSuite.scala

```diff
@@ -0,0 +1,717 @@
+        table.createTransactionBuilder(engine, testEngineInfo, Operation.MANUAL_UPDATE)
+          .withSchema(engine, newSchema)
+          .build(engine).commit(engine, emptyIterable())
+      }
+
+      assert(e.getMessage.contains(
+        "Existing field with id 2 in " +
+          "current schema has nested field IDs [3, 4] which is different from [5, 6]"))
+    }
+  }
+
+  test("Updating schema if map columns are missing nested ids fails") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val schema = new StructType()
+        .add("a", IntegerType.INTEGER, true)
+
+      createTxn(
+        engine,
+        tablePath,
+        isNewTable = true,
+        schema,
+        partCols = Seq.empty,
+        tableProperties = Map(
+          TableConfig.COLUMN_MAPPING_MODE.getKey -> "id",
+          TableConfig.ICEBERG_COMPAT_V2_ENABLED.getKey -> "true")).commit(engine, emptyIterable())
+
+      val currentSchema = table.getLatestSnapshot(engine).getSchema
+      val newSchema = new StructType()
+        .add("a", StringType.STRING, true, currentSchema.get("a").getMetadata)
+        .add(
+          "c",
+          new MapType(
+            StringType.STRING,
+            StringType.STRING,
+            false),
+          true,
+          fieldMetadataForColumn(2, "c"))
+
+      val e = intercept[IllegalArgumentException] {
+        table.createTransactionBuilder(engine, testEngineInfo, Operation.MANUAL_UPDATE)
+          .withSchema(engine, newSchema)
+          .build(engine).commit(engine, emptyIterable())
+      }
+
+      assert(e.getMessage.contains(
+        "Map field c must have exactly 2 nested IDs"))
+    }
+  }
+
+  test("Updating schema if array columns is missing nested id fails") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val schema = new StructType()
+        .add("a", IntegerType.INTEGER, true)
+
+      createTxn(
+        engine,
+        tablePath,
+        isNewTable = true,
+        schema,
+        partCols = Seq.empty,
+        tableProperties = Map(
+          TableConfig.COLUMN_MAPPING_MODE.getKey -> "id",
+          TableConfig.ICEBERG_COMPAT_V2_ENABLED.getKey -> "true")).commit(engine, emptyIterable())
+
+      val currentSchema = table.getLatestSnapshot(engine).getSchema
+      val newSchema = new StructType()
+        .add("a", StringType.STRING, true, currentSchema.get("a").getMetadata)
+        .add("c", new ArrayType(StringType.STRING, false), true, fieldMetadataForColumn(2, "c"))
+
+      val e = intercept[IllegalArgumentException] {
+        table.createTransactionBuilder(engine, testEngineInfo, Operation.MANUAL_UPDATE)
+          .withSchema(engine, newSchema)
+          .build(engine).commit(engine, emptyIterable())
+      }
+
+      assert(e.getMessage.contains(
+        "Array field c must have exactly 1 nested ID"))
+    }
+  }
+
+  def fieldMetadataForColumn(
```

Comment:
just curious: would these be useful as utilities in ColumnMapping? if the answer is "no" --> all good :)

Reply:
Possibly? I'm hard pressed to think that this generalizes well, I feel like it's only useful for this test suite. If I'm wrong, we can always add it later when needed